pkg/apis/admissionregistration/types.go

```diff
@@ -115,6 +115,244 @@ const (
+	//
+	// The `apiVersion`, `kind`, `metadata.name` and `metadata.generateName` are always accessible from the root of the
+	// object. No other metadata properties are accessible.
+	//
+	// Only property names of the form `[a-zA-Z_.-/][a-zA-Z0-9_.-/]*` are accessible.
+	// Accessible property names are escaped according to the following rules when accessed in the expression:
+	// - '__' escapes to '__underscores__'
+	// - '.' escapes to '__dot__'
+	// - '-' escapes to '__dash__'
+	// - '/' escapes to '__slash__'
+	// - Property names that exactly match a CEL RESERVED keyword escape to '__{keyword}__'. The keywords are:
+	//	  "true", "false", "null", "in", "as", "break", "const", "continue", "else", "for", "function", "if",
+	//	  "import", "let", "loop", "package", "namespace", "return".
+	// Examples:
+	//   - Expression accessing a property named "namespace": {"Expression": "object.__namespace__ > 0"}
+	//   - Expression accessing a property named "x-prop": {"Expression": "object.x__dash__prop > 0"}
+	//   - Expression accessing a property named "redact__d": {"Expression": "object.redact__underscores__d > 0"}
+	//
+	// Equality on arrays with list type of 'set' or 'map' ignores element order, i.e. [1, 2] == [2, 1].
+	// Concatenation on arrays with x-kubernetes-list-type use the semantics of the list type:
+	//   - 'set': `X + Y` performs a union where the array positions of all elements in `X` are preserved and
+	//     non-intersecting elements in `Y` are appended, retaining their partial order.
+	//   - 'map': `X + Y` performs a merge where the array positions of all keys in `X` are preserved but the values
+	//     are overwritten by values in `Y` when the key sets of `X` and `Y` intersect. Elements in `Y` with
+	//     non-intersecting keys are appended, retaining their partial order.
+	// Required.
+	Expression string
+	// Message represents the message displayed when validation fails. The message is required if the Expression contains
+	// line breaks. The message must not contain line breaks.
+	// If unset, the message is "failed rule: {Rule}".
+	// e.g. "must be a URL with the host matching spec.host"
+	// If ExpressMessage is specified, Message will be ignored
+	// If the Expression contains line breaks. Eith Message or ExpressMessage is required.
+	// The message must not contain line breaks.
+	// If unset, the message is "failed Expression: {Expression}".
+	// +optional
+	Message string
+	// If the MessageExpression evaluates to an error, the Message field is used to provide the message.
+	// The MessageExpression retured when failed the validation of specified.
+	// If MessageExpression is specified, Message will be ignored
+	// If the Expression contains line breaks. Either Message or MessageExpression is required.
+	// The MessageExpression must not contain line breaks.
+	// +optional
+	MessageExpression string
+	// Reason returned when failed the validation.
+	// +optional
+	Reason string
```

Comment:
How about this?

I've reviewed errors.go and the 400 & 500 HTTP status codes and this looks right to me:

401:

- Unauthorized (since userInfo is available in AdmissionReview and we have authz plans)
- 
403: We expose all the errors.go error types that apply generally (403 is defined as "The request contained valid data and was understood by the server, but the server is refusing action. This may be due to the user not having the necessary permissions for a resource or needing an account of some sort, or attempting a prohibited action...)

- ValueNotFound (to disambiguate from 404 Not found)
- Required
- Duplicate
- Invalid
- Forbidden
- TooLong
- TooMany

500:

- InternalError

I'd be interested if anyone has others to add.

Reply:
Good point. Maybe keep `Forbidden` as 403, make the rest 422?

Reply:
something like that seems reasonable. fwiw, the error constructors in https://github.com/kubernetes/kubernetes/blob/215f236a6b9087c63f4f65fd34a066e56fae54d7/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go and the code → reason mappings in https://github.com/kubernetes/kubernetes/blob/215f236a6b9087c63f4f65fd34a066e56fae54d7/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L435 are good references for expected kubernetes mappings

Reply:
Have updated to use metav1.Status for reason if that sounds ok.

Reply:
Status has a bunch of fields I don't want to expose here. But metav1.StatusReason looks like a good enum to use here.